### ISSUE TITLE
create data reader objects

### ISSUE THREAD

**Issue by CCThompson82** [CCThompson82/nuclei#3]
* object should ease the iteration of data for general applications
* Explore tensorflow data objects instead of python generators or custom built classes

**Comment by CCThompson82** [comment]
* [creating an image TF dataset](https://kratzert.github.io/2017/06/15/example-of-tensorflows-new-input-pipeline.html)

**Comment by CCThompson82** [comment]
* [how to convert to TFRecord data](https://github.com/tensorflow/tensorflow/blob/r1.5/tensorflow/examples/how_tos/reading_data/convert_to_records.py)

**Comment by CCThompson82** [comment]
* [blog post on creating TFRecord data](http://warmspringwinds.github.io/tensorflow/tf-slim/2016/12/21/tfrecords-guide/)

**Comment by CCThompson82** [comment]
* [using queues](https://blog.metaflow.fr/tensorflow-how-to-optimise-your-input-pipeline-with-queues-and-multi-threading-e7c3874157e0)

**Comment by CCThompson82** [comment]
## Plan for Dataset and Reader Construction 

- [ ]  create a tensorflow Dataset object composed of the following elements:
    -  filename to image 
    -  raw label for image 
    -  unique example id 
    -  train/validate/test category?
- [ ] write this dataset permanently to data/interim 
- [ ] utils that can load, display, etc the images and labels from the example_id 

## Notes 
-  map functions will load, preprocess data as it is iterated 
-  intention is to use queue system to speed up training / prediction, as opposed to using binary TFRecord formatted data.

**Comment by CCThompson82** [comment]
## Comments on `Plan for Dataset and Reader Construction` 

After initial attempts at the plan above, several problems were encountered.  First, it seemed that tf.data.Dataset objects are not writable, at least as far as I could figure out.  Without this ability, different training and validation sets will be generated each time a model is run, which is not what I intended.  I next moved to setting up a tfrecord dataset, which I am able to do.  However, this requires each image to be of a fixed size.  The data provided are of various sizes and without understanding the differences among the dataset, I do not want to squash all images to 256x256 in order to create the binary file.

## Updated Plan
* create and write immutable dictionary as as json file for each of the desired datasets.  The object will contain the following keys for each example:
  * imageid
  * full path to the image filename
  * path to the masks for the image (if a train or validation, None if for testing)
* create a generator that will provide one or batches of annotations 
* create utility functions that compiles the individual masks into a single binary mask for the image, given the directory where the masks image files exist.  
* use queue running system to preload and process images and label masks using these utility functions and the generator / reader object.
